SmartDeviceLinkTests/ProtocolSpecs/SDLProtocolReceivedMessageProcessorSpec.m

```diff
@@ -0,0 +1,554 @@
+//
+//  SDLProtocolReceivedMessageProcessorSpec.m
+//  SmartDeviceLinkTests
+//
+//  Created by George Miller on 8/9/22.
+//  Copyright © 2022 smartdevicelink. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+#import <Quick/Quick.h>
+#import <Nimble/Nimble.h>
+
+#import "SDLProtocol.h"
+#import "SDLProtocolReceivedMessageProcessor.h"
+#import "SDLV2ProtocolHeader.h"
+#import "SDLV2ProtocolMessage.h"
+
+typedef NS_ENUM(NSInteger, ProcessorState) {
+    START_STATE = 0x0,
+    SERVICE_TYPE_STATE = 0x01,
+    CONTROL_FRAME_INFO_STATE = 0x02,
+    SESSION_ID_STATE = 0x03,
+    DATA_SIZE_1_STATE = 0x04,
+    DATA_SIZE_2_STATE = 0x05,
+    DATA_SIZE_3_STATE = 0x06,
+    DATA_SIZE_4_STATE = 0x07,
+    MESSAGE_1_STATE = 0x08,
+    MESSAGE_2_STATE = 0x09,
+    MESSAGE_3_STATE = 0x0A,
+    MESSAGE_4_STATE = 0x0B,
+    DATA_PUMP_STATE = 0x0C,
+    ERROR_STATE = -1,
+};
+
+@interface SDLProtocolReceivedMessageProcessor(){}
+// State management
+@property (assign, nonatomic) ProcessorState state;
+
+// Message assembly state
+@property (strong, nonatomic) SDLProtocolHeader *header;
+@property (strong, nonatomic) NSMutableData *headerBuffer;
+@property (strong, nonatomic) NSMutableData *payloadBuffer;
+
+// Error checking
+@property (assign, nonatomic) UInt8 version;
+@property (assign, nonatomic) BOOL encrypted;
+@property (assign, nonatomic) SDLFrameType frameType;
+@property (assign, nonatomic) UInt32 dataLength;
+@property (assign, nonatomic) UInt32 dataBytesRemaining;
+@property (assign, nonatomic) SDLServiceType serviceType;
+@end
+
+QuickSpecBegin(SDLProtocolReceivedMessageProcessorSpec)
+
+describe(@"The processor", ^{
+    __block SDLProtocolReceivedMessageProcessor *testProcessor = nil;
+    __block NSMutableData *testBuffer;
+    __block NSMutableData *testHeaderBuffer;
+    
+    beforeEach(^{
+        testProcessor = [[SDLProtocolReceivedMessageProcessor alloc] init];
+        testBuffer = [NSMutableData data];
+        testHeaderBuffer = [NSMutableData data];
+    });
+    context(@"in START_STATE", ^{
+        beforeEach(^{
+            testProcessor.state = START_STATE;
+        });
+        context(@"When it recieves a byte", ^{
+            context(@"with a good version 1", ^{
+                it(@"transitions to SERVICE_TYPE_STATE", ^{
+                    Byte testByte = 0x11;
+                    [testBuffer appendBytes:&testByte length:1];
+                    
+                    [testProcessor processReceiveBuffer:testBuffer withMessageReadyBlock:^(SDLProtocolHeader *header, NSData *payload) {
+                        //do nothing?
+                    }];
+                    expect(@(testProcessor.state)).to(equal(SERVICE_TYPE_STATE));
+                });
+            });
+            context(@"with a good version 2", ^{
+                it(@"transitions to SERVICE_TYPE_STATE", ^{
```

Comment:
I think you should restructure slightly. Do one `context` "with a good version byte", then multiple `it`s within "version 1 transition to...", etc.

Reply:
I'm reading up on the structure because this isn't making sense to me.
I read "describe" and thing "description of a test".
I read "context" and I think of the variables that are critical to a test.
I took "it" to mean the expected behavior of the test.
But in that way, I can't have multiple "it"s because the critical variables that make a test unique are communicated in the context.

What I'm finding is that the nimble documentation is no help, and the quick documentation uses test classes.  Do we have a document that talks about out protocol for using quick and simple to setup a test?